include/kernel.h

```diff
@@ -3251,12 +3278,14 @@ extern void k_work_q_start(struct k_work_q *work_q,
  *		should either be the same constant passed to
  *		K_THREAD_STACK_DEFINE() or the value of K_THREAD_STACK_SIZEOF().
  * @param prio Priority of the work queue's thread.
+ * @param options Option flags to be used in the resulting work_q k_thread's
+ * creation.
  *
  * @return N/A
+ * @req K-WORK-001
  */
-extern void k_work_q_user_start(struct k_work_q *work_q,
-				k_thread_stack_t *stack,
-				size_t stack_size, int prio);
+extern void k_work_q_user_start_ex(struct k_work_q *work_q, k_thread_stack_t *stack,
+				size_t stack_size, int prio, u32_t options);
```

Comment:
Stylewise: I'm not a fan of this kind of API name extension (c.f. wait, waitpid, wait3, wait4... Win32Function, Win32FunctionEx...).   We've literally already done this on this very call: what used to be k_work_q_start() became two different calls, and now it's become four!

Personally I'd prefer to see a scheme that introduces the new "correct" API and deprecates the old one(s).

How about: "I've created a thread of my own via some path you don't care about and now it's running.  Please enter the main loop for this work queue."

Reply:
Seems a single API could cover all the use cases:

```c
void k_work_q_start_new(struct k_work_q *work_q, k_thread_stack_t *stack, size_t stack_size, int prio, u32_t options)
{
    /* ... */

    if (options & K_USER) {
        /* do usermode stuff here. */
    }

   /* ... */
}
```

Reply:
FWIW: my proposal was to leave the creation of the thread and its entry point under the control of the user and make the work_q's main loop a plain function to be called from any thread and not a special thing.  But I'm not that picky.  I just think we need to draw the line at "fewer than four variants of a single function"

Reply:
> FWIW: my proposal was to leave the creation of the thread and its entry point under the control of the user and make the work_q's main loop a plain function to be called from any thread and not a special thing. But I'm not that picky. I just think we need to draw the line at "fewer than four variants of a single function"

Ah I understand, I'll admit that's a pretty clean solution. I think we'd just need to change z_work_q_main to a global function (k_work_q_main?), update documentation, and potentially deprecate the old APIs.